cpp/src/reductions/reduction_operators.cuh

```diff
@@ -0,0 +1,198 @@
+
+#pragma once
+
+#include <utilities/cudf_utils.h>   //for CUDA_HOST_DEVICE_CALLABLE
+#include <cudf/detail/iterator.cuh>
+#include <iterator/transform_unary_functions.cuh>
+#include <utilities/device_operators.cuh>
+
+#include <cmath>
+#include <thrust/functional.h>
+
+namespace cudf {
+namespace experimental {
+namespace reduction {
+
+// intermediate data structure to compute `var`, `std`
+template<typename ResultType>
+struct var_std
+{
+    ResultType value;                /// the value
+    ResultType value_squared;        /// the value of squared
+
+    CUDA_HOST_DEVICE_CALLABLE
+    var_std(ResultType _value=0, ResultType _value_squared=0)
+    : value(_value), value_squared(_value_squared)
+    {};
+
+    using this_t = var_std<ResultType>;
+
+    CUDA_HOST_DEVICE_CALLABLE
+    this_t operator+(this_t const &rhs) const
+    {
+        return this_t(
+            (this->value + rhs.value),
+            (this->value_squared + rhs.value_squared)
+        );
+    };
+};
+
+// transformer for `struct var_std` in order to compute `var`, `std`
+template<typename ResultType>
+struct transformer_var_std
+{
+    using OutputType = var_std<ResultType>;
+
+    CUDA_HOST_DEVICE_CALLABLE
+    OutputType operator() (ResultType const & value)
+    {
+        return OutputType(value, value*value);
+    };
+};
+
+// ------------------------------------------------------------------------
+// difinitions of device struct for reduction operation
+// all `op::xxx` must have `Op` and `make_iterator`
+// `Op`  is used to compute the reduction at device
+// `make_iterator` is used to make a iterator to be used to compute the reduction at device
+
+namespace op {
+
+// `product`, `product`, `sum_of_squares`, `min`, `max`
+// are used at simple_reduction
+
+struct sum {
+    using Op = cudf::DeviceSum;
+
+    template<typename ResultType>
+    using transformer = thrust::identity<ResultType>;
+};
+
+struct product {
+    using Op = cudf::DeviceProduct;
+
+    template<typename ResultType>
+    using transformer = thrust::identity<ResultType>;
+};
+
+struct sum_of_squares {
+    using Op = cudf::DeviceSum;
+
+    template<typename ResultType>
+    using transformer = cudf::transformer_squared<ResultType>;
+};
+
+struct min {
+    using Op = cudf::DeviceMin;
+
+    template<typename ResultType>
+    using transformer = thrust::identity<ResultType>;
+
+};
+
+struct max {
+    using Op = cudf::DeviceMax;
+
+    template<typename ResultType>
+    using transformer = thrust::identity<ResultType>;
+};
+
+
+// `mean`, `variance`, `standard_deviation` are used at compound_reduction
+// compound_reduction requires intermediate::IntermediateType and intermediate::compute_result
+// IntermediateType is the intermediate data structure type of a single reduction call,
+// it is also used as OutputType of cudf::reduction::detail::reduce at compound_reduction.
+// compute_result computes the final ResultType from the IntermediateType.
+
+// operator for `mean`
+struct mean {
```

Comment:
All of these compound operators are expected to hold to a common interface (e.g., define an `intermediate` type, define `compute_results`, etc.), but there isn't anything that actually enforces that common interface. Seems like there should be a `compound_reduction` base class that defines what the interface is and have these compound reduction types derive from it.

Reply:
enforce these template functions with CRTP?

Reply:
implemented `CompoundOp` CRTP and enforced on compound operators.
But SFINAE with std::is_base_of does not compile in `detail::reduce()` method.

Reply:
CRTP is good choice, but needs some work.